Add `cloudflare_zone` data resource (singular)

cmackenzie1 opened the pull request:
This commit adds `cloudflare_zone` as a data source. While support for this technically exists by using `cloudflare_zones` and a `lookup`, I think we should explicitly support this approach as well for providing data about zones.

One concern I have is that the same set of fields should be returned when using `cloudflare_zone` or `cloudflare_zones`. Currently, `cloudflare_zones` only returns `id` and `name`.

jacobbednarz commented:
Appreciate the effort here @cmackenzie1! 🙇 

I've been pretty reluctant to add a singular resource for zones largely because many people aren't aware that the zone name attribute isn't unique within Cloudflare. Meaning, if I have 5 accounts, it is possible to have 5 zones for "example.com" within each account, and it's totally acceptable (in fact some customers intentionally do this for migrations!). This makes a singular resource very difficult to use correctly and with any confidence. 

Using my example of 5 accounts with "example.com" and a singular resource, the configuration could look like this:

```hcl
data "cloudflare_zone" "my_zone" {
  name = "example.com"
}
```

Immediately, a bunch of questions come to mind. Which zone ID should be returned here? Do we introduce ordering to weight the zone status higher and just take that? How do we handle similar cases? Do we just error out and advise to be more specific with your lookup? There _are_solutions to some of these questions such as adding filtering to the resource and/or including the account ID in the resource but at that point, we're pretty much using the same functionality in the pluralised data source and the only difference is the 3-4 characters that are a part of the index access. 

At this point, I'm still struggling to see a valid use case for introducing and supporting this as a singular resource given the small improvements we make over the pluralised version. I'll sit on this for another few days and see what I can come up with and whether we introduce this.

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+				Computed: true,
+			},
+			"status": {
+				Type:     schema.TypeBool,
```
`status` is a string in the [zone struct](https://github.com/cloudflare/cloudflare-go/blob/master/zone.go#L42)

```suggestion
				Type:     schema.TypeString,
```

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+		Read: dataSourceCloudflareZoneRead,
+
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"account_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
```
looking at the use cases, i don't think it makes sense to ever apply these two together. if you want clarity, zone ID is unique and should be enough. what was your use case here?

```suggestion
			"zone_id": {
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
				ConflictsWith: []string{"name"},
			},
			"account_id": {
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
			},
			"name": {
				Type:     schema.TypeString,
				Optional: true,
				Computed: true,
				ConflictsWith: []string{"zone_id"},
			},
```

cmackenzie1 replied:
Oh sweet! I didn't know there was a `ConflictsWith` option

jacobbednarz replied:
np! check out everything that is available at https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/schema.go#L196 if you're curious what else a schema can hold

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+	name = name.(string)
+
+	if nameExists && zoneIDExists {
+		return fmt.Errorf("zone_id and name arguments can't be used together")
```
heh, you're actually already preventing zone ID and name from being used together like i suggested further down. might just need to update the documentation to reflect this.

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		if len(zones) >= 1 {
+			return fmt.Errorf("more than one zone was returned")
```
```suggestion
			return fmt.Errorf("more than one zone was returned; you should consider using `cloudflare_zones` data source with filtering to target the zone more specifically")
```

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+			return fmt.Errorf("error listing zones: %s", err)
+		}
+
+		if len(zones) >= 1 {
```
the logic here is a little off; having a single record is 👌 

```suggestion
		if len(zones) > 1 {
```

jacobbednarz commented on cloudflare/data_source_zone.go:
```diff
@@ -0,0 +1,109 @@
+
+	var zone cloudflare.Zone
+	if nameExists && !zoneIDExists { // if only the name was provided
+		zones, err := client.ListZones(context.Background(), name.(string))
```
alternatively, you could use `ListZonesContext` here and pass in `WithZoneFilters` with the zone name and account ID if it is set. example: https://github.com/cloudflare/cloudflare-go/blob/b8935550e028d7c46a7adb20ab8c48f8e849bd06/cloudflare.go#L146

cmackenzie1 commented:
Thanks for the feedback and response @jacobbednarz 🙏! Until you pointed it out I thought zone's were globally unique on Cloudflare.

I incorporated some of the feedback you left here and added the ability to specify an `accountID` to be used when querying for zones. 

I'll hold off on writing tests and such until you decide if singular resource data definitions should be added or not.

jacobbednarz commented:
Sorry for the delay in coming back to this @cmackenzie1. I baked on this over the weekend and got caught up with some other reviews. I'm still not fully sold on this as a data source because of the reasoning I mentioned above but at the same time, we have an alternative if people hit those issues _and_ we've made it clear in the documentation there is a potential for getting the wrong data here. 

Let's land this and see how it goes. Thanks for getting the PR together!
